Optimize for small numbers of arguments

dead-claudia opened the pull request:
I also have a tagalong commit to expand the README to include license/examples (and reference the related [ES function bind proposal](https://github.com/zenparsing/es-function-bind)), but it's easy to remove, if you would rather leave that out.

Commit:
Optimize for small numbers of arguments

Commit:
Expand README to include license/examples

barneycarroll commented:
Thanks @isiahmeadows! Without PC until Monday so won't be merging until then. 

Not sure I understand the arbitrary number of 'N arguments' tests. I don't believe these tests actually prove or clarify anything that wasn't tested for and obvious before — I can see how they might convince somebody with a poor grasp of numeracy, but I don't want to pander to that kind of thinking. I would rather deal with that confusion as and when somebody actually suffers from it — a bit like that 'active tree' thing with Mithril, I think it's a mistake to pre-empt those kinds of concerns.

dead-claudia commented:
@barneycarroll Just in case. Typos and dumb mistakes happen, so IMHO too many tests are better than not enough.

The reason I added them is because of the special cases I added. 0, 1, 2, 3, and 4 are all special-cased, while 5 and greater are not. Each special case has its own corresponding test, and the general case has two: 5 to confirm it does its job at the lower bound, and 10 to just make sure it doesn't mysteriously break with anything higher. Call me cautious, if you would like.

barneycarroll commented:
Right, I get it. Reading this on a mobile didn't quite make the click :)

Still, all this seems arbitrary to an outsider — I basically have to trust the wisdom of your assertion that this special casing happens in implementing engines and is worth pursuing. Can we add tests to validate this? If not, it's basically adding a lot of code with no internal reasoning.

dead-claudia commented on asarg.js:
```diff
@@ -1,10 +1,35 @@
+				arguments[ 3 ] )
+
+		default:
+			var subject = this
```
Oops...will fix :smile:

dead-claudia commented:
@barneycarroll Trust me in that it's not quite arbitrary: [Node's process.nextTick](https://github.com/nodejs/node/blob/master/src/node.js#L427-L429), [Node's EventEmitter::emit](https://github.com/nodejs/node/blob/master/lib/events.js#L74-L78), [Bluebird on calling methods](https://github.com/petkaantonov/bluebird/blob/7a39370ba1b98da0aaef1fa9d85b2fd5daaba4ee/src/call_get.js#L17-L33), [Bluebird on promisifying in Node (which also uses code generation)](https://github.com/petkaantonov/bluebird/blob/468fa51ab6aa6f422f55cc9c141a57564bc74c73/src/promisify.js#L91), and Lodash in [function application](https://github.com/lodash/lodash/blob/master/lodash.js#L401-L420), [constructor application](https://github.com/lodash/lodash/blob/master/lodash.js#L4158-L4189), and [rest param application](https://github.com/lodash/lodash/blob/master/lodash.js#L9141-L9152).

Most modern engines (even IE 10) optimize for a consistent number of arguments. As long as this holds, it can transparently reduce the function call to a dozen or so assembly instructions at most without any detectable changes other than speed.

In the case of V8 specifically, it literally allocates a new C++ array unless the second argument is null/undefined/not given. It's more optimal with frequently varying numbers of arguments than with a constant number of them. [1](https://github.com/v8/v8/blob/c0198937d8a2a78b358fe6096e35a245d26b3a37/src/ia32/builtins-ia32.cc#L1081), [2](https://github.com/v8/v8/blob/master/src/runtime/runtime-function.cc#L289), [3](https://github.com/v8/v8/blob/master/src/execution.cc#L123

As for this module, most functions this creates are going to be called with the same number of arguments constantly. This dispatch method will allow engines to optimize for the same number of arguments for these functions. I'm going to correct a few small things, though.

dead-claudia commented:
Most of my remaining corrections are just little things.

dead-claudia commented:
Sadly, the only way to test that engines optimize that is by speed
comparison. Native calls can't even do the trick here, since the
optimization is technically an implementation detail with no other
observable difference in every engine I know of.

(It takes advantage of polymorphic inline caches commonly used to optimize
for consistent argument types, and these aren't usually accessible in any
way outside the internals of the engine.)

As for the other difference between this and the old implementation, it's
also removing a layer of indirection for the common cases by invoking the
function directly instead of creating a JS array in the middle. That alone
would cut execution time down to a fraction. I think this reason is also a
little more accessible (without going into engine internals).

On Mon, Feb 15, 2016, 04:22 Barney Carroll notifications@github.com wrote:

> Right, I get it. Reading this on a mobile didn't quite make the click :)
> 
> Still, all this seems arbitrary to an outsider — I basically have to trust
> the wisdom of your assertion that this special casing happens in
> implementing engines and is worth pursuing. Can we add tests to validate
> this? If not, it's basically adding a lot of code with no internal
> reasoning.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/barneycarroll/asarg/pull/2#issuecomment-184128732.
